### ISSUE TITLE
refactor(jupyter): deduplicate post-pip jupyter provisioning RUN blocks

### ISSUE THREAD

**Issue by ide-developer** [opendatahub-io/notebooks#3940]
## Summary

Leaf Jupyter workbench Dockerfiles repeat the same **post-`uv pip install` provisioning** heredoc: Elyra/runtime cleanup, kernel `display_name` tweak, `jupyter_server_config.py` install, `jupyter labextension disable`, `labconfig` `chmod`, `apply.sh`, and `ensure-openshift-site-packages.sh`.

The block is copy-pasted across **10** `Dockerfile.konflux.*` leaf images with minor ordering/comment drift. Recent OpenShift permission work ([#3928](https://github.com/opendatahub-io/notebooks/issues/3928) / [#3934](https://github.com/opendatahub-io/notebooks/pull/3934)) had to touch every copy (e.g. `install -m 0664`, `chmod g+w labconfig`, `USER 1001:0`), which is brittle and already caused a regression in **trustyai** where the second `RUN` dropped back to `USER 1001` instead of `USER 1001:0`.

This issue is **not** covered by the existing Dockerfile dedup plan ([#3355](https://github.com/opendatahub-io/notebooks/issues/3355) / [RHOAIENG-54488](https://issues.redhat.com/browse/RHOAIENG-54488)), which targets Konflux twin files and cpu/cuda/rocm **file** merges — not shared **RUN** content inside those files.

## Affected files (today)

All contain a near-identical block after pip install:

- `jupyter/minimal/ubi9-python-3.12/Dockerfile.konflux.{cpu,cuda,rocm}`
- `jupyter/datascience/ubi9-python-3.12/Dockerfile.konflux.cpu`
- `jupyter/trustyai/ubi9-python-3.12/Dockerfile.konflux.cpu` *(split across two `RUN`s — extra footgun)*
- `jupyter/pytorch/ubi9-python-3.12/Dockerfile.konflux.cuda`
- `jupyter/pytorch+llmcompressor/ubi9-python-3.12/Dockerfile.konflux.cuda`
- `jupyter/tensorflow/ubi9-python-3.12/Dockerfile.konflux.cuda`
- `jupyter/rocm/pytorch/ubi9-python-3.12/Dockerfile.konflux.rocm`
- `jupyter/rocm/tensorflow/ubi9-python-3.12/Dockerfile.konflux.rocm` *(also has a **later** site-packages `COPY` after ensure — see [#3934](https://github.com/opendatahub-io/notebooks/pull/3934) review)*

Representative duplicated steps:

```dockerfile
mkdir /opt/app-root/runtimes
rm /opt/app-root/share/jupyter/metadata/runtime-images/*.json
sed -i -e "s/Python.*/$(python --version | cut -d '.' -f-2)\",/" \
  /opt/app-root/share/jupyter/kernels/python3/kernel.json
install -D -m 0664 /opt/app-root/bin/utils/jupyter_server_config.py \
  /opt/app-root/etc/jupyter/jupyter_server_config.py
jupyter labextension disable "@jupyterlab/apputils-extension:announcements"
jupyter labextension disable jupyterlab-kubeflow-kale   # most leaves; minimal omits kale
chmod g+w /opt/app-root/etc/jupyter/labconfig
/opt/app-root/bin/utils/addons/apply.sh
/usr/local/bin/ensure-openshift-site-packages.sh
```

## Variants to parameterize

| Difference | Examples |
|------------|----------|
| `pipeline-runtimes` mkdir | datascience only |
| Kale disable | most leaves; minimal skips |
| Split vs single `RUN` | trustyai (pip in RUN 1, jupyter setup in RUN 2) |
| Post-ensure site-packages `COPY` | rocm tensorflow (`usercustomize.pth`, `monkey_patch_protobuf_6x.py`) |
| Step order | minimal reorders labextension vs sed vs install |

## Proposed direction

Pick one (or combine):

1. **`base-images/utils/setup-jupyter-workbench.sh`** — invoked once from each leaf `RUN` under `USER 1001:0` + `umask 0002`; flags/env for kale, pipeline-runtimes, etc.
2. **`scripts/dockerfile_fragments.py` marker** — extend the existing `### BEGIN` / `### END` sync mechanism ([`scripts/dockerfile_fragments.py`](scripts/dockerfile_fragments.py), documented in [`scripts/README.md`](scripts/README.md#dockerfile_fragmentspy), tested under [#2683](https://github.com/opendatahub-io/notebooks/issues/2683)) so the heredoc body is generated in all leaf Dockerfiles from one source.

Either approach should preserve the OpenShift contract documented in [#3928](https://github.com/opendatahub-io/notebooks/issues/3928) / [`ARCHITECTURE.md`](ARCHITECTURE.md): `USER 1001:0`, `umask 0002`, surgical modes at creation, `ensure-openshift-site-packages.sh` last among site-packages mutations.

## Related issues / docs

| Ref | Relevance |
|-----|-----------|
| [#3355](https://github.com/opendatahub-io/notebooks/issues/3355) | Umbrella Dockerfile dedup (Konflux twins + variant merge); complementary, not a substitute |
| [RHOAIENG-54488](https://issues.redhat.com/browse/RHOAIENG-54488) | Jira twin of #3355 ([`docs/konflux.md`](docs/konflux.md#dockerfile-deduplication)) |
| [#3928](https://github.com/opendatahub-io/notebooks/issues/3928) | Permission model this block must keep |
| [#1321](https://github.com/opendatahub-io/notebooks/issues/1321) | Replace fragile kernel `sed` with JSON-safe Python while deduplicating |
| [#2683](https://github.com/opendatahub-io/notebooks/issues/2683) | Stronger tests for `dockerfile_fragments.py` if we use markers |
| [`scripts/dockerfile_fragments.py`](scripts/dockerfile_fragments.py) | Existing codegen for PDF export deps, pip/uv config, subscription-manager, etc. |

## Acceptance criteria

- [ ] One authoritative definition of the post-pip jupyter provisioning steps
- [ ] All 10 leaf `Dockerfile.konflux.*` call it (trustyai merged back to single `RUN` or script covers both)
- [ ] OpenShift permission battery from #3928 still passes (site-packages + etc/share jupyter g+w, arbitrary UID pip)
- [ ] #1321 kernel JSON fix can land in the shared location without touching 10 files
- [ ] `make test` / alignment checks green

## Non-goals

- Merging cpu/cuda/rocm Dockerfiles ([#3355](https://github.com/opendatahub-io/notebooks/issues/3355) phase 2) — can be done before or after this
- Changing Konflux LABEL parameterization ([#3355](https://github.com/opendatahub-io/notebooks/issues/3355) phase 1)

**Comment by ide-developer** [comment]
@coderabbitai Please review this issue and recommend the best deduplication approach for the post-pip Jupyter provisioning block.

**Context:** While landing OpenShift permission fixes ([#3928](https://github.com/opendatahub-io/notebooks/issues/3928) / [#3934](https://github.com/opendatahub-io/notebooks/pull/3934)), we had to edit the same ~15-line heredoc in **10** leaf `Dockerfile.konflux.*` files. That is separate from the file-level dedup tracked in [#3355](https://github.com/opendatahub-io/notebooks/issues/3355) / [RHOAIENG-54488](https://issues.redhat.com/browse/RHOAIENG-54488).

**Questions for triage:**

1. **`setup-jupyter-workbench.sh`** (runtime script copied into the image) vs **`dockerfile_fragments.py` marker** (build-time sync) — which fits this repo better given existing patterns in [`scripts/dockerfile_fragments.py`](scripts/dockerfile_fragments.py)?
2. How should we handle **per-image variants** (datascience `pipeline-runtimes`, kale disable, trustyai split `RUN`, rocm tensorflow post-ensure `COPY`) without reintroducing drift?
3. Should [#1321](https://github.com/opendatahub-io/notebooks/issues/1321) (kernel JSON `sed` → Python) be folded into the same shared location?
4. Any ordering constraints with `ensure-openshift-site-packages.sh` and later site-packages `COPY`s that the shared abstraction must enforce?

Please flag if this should be a sub-task of #3355 or kept standalone.

**Comment by jiridanek** [comment]
## Trajectory

**Verdict:** SKIP (design decision required — not a deterministic bug fix)

### Verification on `origin/main` (`f3f32101b`)

The duplicated post-`uv pip install` provisioning block **still exists** in all 11 `Dockerfile.konflux.*` leaf files at `origin/main` (confirmed via `git show origin/main:<file>` + `grep -c ensure-openshift-site-packages.sh`):

| File | `ensure-openshift-site-packages.sh` occurrences |
|------|------|
| `jupyter/baseline/ubi9-python-3.12/Dockerfile.konflux.cpu` | 2 |
| `jupyter/datascience/ubi9-python-3.12/Dockerfile.konflux.cpu` | 3 |
| `jupyter/minimal/ubi9-python-3.12/Dockerfile.konflux.{cpu,cuda,rocm}` | 2 each |
| `jupyter/pytorch/ubi9-python-3.12/Dockerfile.konflux.cuda` | 2 |
| `jupyter/pytorch+llmcompressor/ubi9-python-3.12/Dockerfile.konflux.cuda` | 2 |
| `jupyter/tensorflow/ubi9-python-3.12/Dockerfile.konflux.cuda` | 2 |
| `jupyter/rocm/pytorch/ubi9-python-3.12/Dockerfile.konflux.rocm` | 2 |
| `jupyter/rocm/tensorflow/ubi9-python-3.12/Dockerfile.konflux.rocm` | 2 |
| `jupyter/trustyai/ubi9-python-3.12/Dockerfile.konflux.cpu` | 3 (split across two `RUN`s) |

So the issue is **not stale** — the duplication is real and present on current main.

### Why this is SKIP rather than IMPLEMENTED

This is an **enhancement/refactoring** issue (labels: `enhancement`, `help wanted`, `priority/minor`, `triaged`), not a bug, and the **design is explicitly unsettled**:

1. **The issue's own comment thread is a design question, not a bug report.** The opening comment (`ide-developer` → `@coderabbitai`) asks for a *recommendation* on which of two proposed approaches to take — (a) a runtime `setup-jupyter-workbench.sh` script, or (b) a new `scripts/dockerfile_fragments.py` marker — plus how to handle per-image variants and ordering constraints. No decision has been recorded in the thread.

2. **The Dockerfiles are generated, so the "correct" fix is a new generator fragment, not 11 hand edits.** `ci/generate_code.sh` runs `uv run scripts/dockerfile_fragments.py` in CI (the `check-generated-code` gate). The generator's `replacements` dict currently manages named fragments (`AIPCC pip and uv config files`, `Dependencies for PDF export`, `mongocli-builder stage`, `Install software and packages`, etc.) via `ntb.blockinfile`. The post-pip jupyter provisioning block is **not** in that dict — adding it is a new feature to the generator, which is a design decision, not a mechanical change.

3. **The variants are non-trivial and must be preserved** (per the issue's own table): datascience-only `mkdir /opt/app-root/runtimes`, trustyai's split `RUN` (pip in RUN 1, jupyter setup in RUN 2), rocm tensorflow's **later** site-packages `COPY` after `ensure-openshift-site-packages.sh`, and per-file step ordering. A blind dedup risks regressing the OpenShift permission contract from #3928 — the issue itself notes a trustyai regression already occurred during the #3934 permission work.

4. **No PR addresses this issue.** Verified via `gh search prs "3940"` (no results) and the issue timeline (only cross-refs to #3941, #3934, #3928).

### Recommendation

Keep open as an enhancement. Before implementation, the maintainers need to pick the approach (runtime script vs. generator marker) and specify how the four variants are parameterized. Once that decision is made, the change is well-scoped: add one named fragment to `scripts/dockerfile_fragments.py` (approach b, preferred since it fits the existing `### BEGIN`/`### END` codegen mechanism and the `check-generated-code` CI gate) and regenerate all 11 files.
